### ISSUE TITLE
reverse-stepping appears to hang forever if no branches occur for a very long time

### ISSUE THREAD

**Issue by Keno** [mozilla/rr#2371]
Consider the following:
```
#include <sys/mman.h>

int
main(int argc, char *argv[])
{
    void *p = mmap(0, 0x100000, PROT_READ | PROT_WRITE | PROT_EXEC, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
    mprotect((char*)p + 0x100000 - 0x1000, 0x1000, PROT_NONE);
    __asm__(
    "movq %%rax, %0\n"
    "jmpq *%%rax" : : "r"(p));
    return 0;
}
```
Running to the inevitable segfault and reverse-stepping takes forever because there is many thousands of instructions between the last tick and the end of the region of zeros that rr tries to PTRACE_SINGLESTEP through. Of course this might be outside the scope of the rr replayer, since instrumentation based replayers will handle this fine, but it'd be nice to fix since bugs like these can happen (in our case the jit wasn't done copying over the generated code on another thread) and are a bit of a reverse-continue trap.

cc @JeffBezanson who hit this issue

**Comment by khuey** [comment]
This is sort of known and in the tool that roc and I have built we actually insert conditional branches that are never taken precisely to avoid this.  You might be able to do something similar in your JIT (conditioned on `RUNNING_UNDER_RR` being set in the environment).

**Comment by Keno** [comment]
I've seen debuggers that have on-board disassemblers scan the instruction stream ahead to find the next branch rather than just singlestepping through. That might be too complex since rr doesn't currently have a full features x86 disassembler, but it would at least solve this case.

**Comment by rocallahan** [comment]
Scanning ahead wouldn't handle the nasty edge case of self-modifying code...

**Comment by Keno** [comment]
Fair point. I guess it'd be useful if either the code is W^X, or none of the instructions write memory, but the latter case is admittedly very tailored to my particular failure example.
